Remove keyword values for mathsize when MathMLCore is enabled

fred-wang opened the pull request:
#### df8ef2e9a52d9261476efa81948c9a38e0507bf7
<pre>
Remove keyword values for mathsize when MathMLCore is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=197493">https://bugs.webkit.org/show_bug.cgi?id=197493</a>

Reviewed by Manuel Rego Casasnovas.

Currently the mathsize attribute relies on convertMathSizeIfNeeded to
convert legacy MathML3 names (`small`, `normal`, `big`) and converts
unitless values to percentages. It is then passed to the CSS parser
which accepts other `font-size` values (such as `x-small`, `larger`,
etc) [1] even if they are not part of MathML.

MathML Core only accepts `&lt;length-percentage&gt;` instead [2]. This commit
adds a new path when MathML Core is enabled. It passes the mathsize value
as it, removing CSS `&lt;absolute-size&gt;`, `&lt;relative-size&gt;` and `math`
keywords. This change is thus removing support for legacy MathML3 names,
nonzero unitless values and CSS keywords. However, due to some weird CSS
bug [3] we still need to explicitly disallow nonzero unitless values.

* LayoutTests/TestExpectations: Remove failures for three WPT tests.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/lengths-2.html: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/lengths-2-expected.txt: Update expectation since we made some progress.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/css-styling/lengths-2-expected.txt: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/mathsize-attribute-css-keywords.html: Force MathML Core to be enabled, so &quot;small&quot; is not allowed.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/mathsize-attribute-legacy-values.html: Ditto.
* LayoutTests/mathml/presentation/attributes-mathsize.html: Force MathML Core to be disabled, so legacy features are still supported.
* Source/WebCore/mathml/MathMLElement.cpp:
(WebCore::isDisallowedMathSizeAttribute): Helper function to remove CSS font-size keywords and nonzero unitless values (i.e. not &lt;length-percentage&gt;).
(WebCore::MathMLElement::collectPresentationalHintsForAttribute): Add new path explained above when MathML Core is enabled.

[1] <a href="https://www.w3.org/TR/css-fonts-4/#font-size-prop">https://www.w3.org/TR/css-fonts-4/#font-size-prop</a>
[2] <a href="https://w3c.github.io/mathml-core/#dfn-mathsize">https://w3c.github.io/mathml-core/#dfn-mathsize</a>
[3] <a href="https://bugs.webkit.org/show_bug.cgi?id=245927">https://bugs.webkit.org/show_bug.cgi?id=245927</a>

Canonical link: <a href="https://commits.webkit.org/255112@main">https://commits.webkit.org/255112@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e8e63f0f26f4aa78123e21974367b579bb5db43

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91410 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/456 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22056 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101136 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161115 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95415 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/471 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83756 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97506 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97068 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27305 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82272 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/81931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70336 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35545 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/15957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33330 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17047 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3558 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37130 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39050 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36158 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

mrego reviewed (approved):
I don't know how to comment on the commit message, so let's try here.

Whey is this `[GTK][WPE]` only?

And about the commit message it looks we're missing something at:
> MathML Core only accepts  instead [2]

Also there are some weird double spacing there and lately in the message. Could you review and update it?

fred-wang committed:
Remove keyword values for mathsize when MathMLCore is enabled
https://bugs.webkit.org/show_bug.cgi?id=197493

Reviewed by Manuel Rego Casasnovas.

Currently the mathsize attribute relies on convertMathSizeIfNeeded to
convert legacy MathML3 names (`small`, `normal`, `big`) and converts
unitless values to percentages. It is then passed to the CSS parser
which accepts other `font-size` values (such as `x-small`, `larger`,
etc) [1] even if they are not part of MathML.

MathML Core only accepts `<length-percentage>` instead [2]. This commit
adds a new path when MathML Core is enabled. It passes the mathsize value
as it, removing CSS `<absolute-size>`, `<relative-size>` and `math`
keywords. This change is thus removing support for legacy MathML3 names,
nonzero unitless values and CSS keywords. However, due to some weird CSS
bug [3] we still need to explicitly disallow nonzero unitless values.

* LayoutTests/TestExpectations: Remove failures for three WPT tests.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/lengths-2.html: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/lengths-2-expected.txt: Update expectation since we made some progress.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/css-styling/lengths-2-expected.txt: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/mathsize-attribute-css-keywords.html: Force MathML Core to be enabled, so "small" is not allowed.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/mathsize-attribute-legacy-values.html: Ditto.
* LayoutTests/mathml/presentation/attributes-mathsize.html: Force MathML Core to be disabled, so legacy features are still supported.
* Source/WebCore/mathml/MathMLElement.cpp:
(WebCore::isDisallowedMathSizeAttribute): Helper function to remove CSS font-size keywords and nonzero unitless values (i.e. not <length-percentage>).
(WebCore::MathMLElement::collectPresentationalHintsForAttribute): Add new path explained above when MathML Core is enabled.

[1] https://www.w3.org/TR/css-fonts-4/#font-size-prop
[2] https://w3c.github.io/mathml-core/#dfn-mathsize
[3] https://bugs.webkit.org/show_bug.cgi?id=245927

Canonical link: https://commits.webkit.org/255112@main
